File: app/src/main/java/name/monwf/customiuizer/prefs/PreferenceCategoryEx.java
package name.monwf.customiuizer.prefs;

import android.content.Context;
import android.content.res.Resources;
import android.content.res.TypedArray;
import android.util.AttributeSet;
import android.view.View;
import android.widget.TextView;

import androidx.preference.Preference;
import androidx.preference.PreferenceCategory;
import androidx.preference.PreferenceViewHolder;

import name.monwf.customiuizer.R;

public class PreferenceCategoryEx extends PreferenceCategory {
	private final boolean dynamic;
	private int state; // 0-正常 1-纯区块 2-顶层隐藏
	private boolean unsupported = false;
	private final Resources res = getContext().getResources();
	private final int childpadding = res.getDimensionPixelSize(R.dimen.preference_item_child_padding);

	public PreferenceCategoryEx(Context context, AttributeSet attrs) {
		super(context, attrs);
		final TypedArray xmlAttrs = context.obtainStyledAttributes(attrs, R.styleable.PreferenceCategoryEx);
		dynamic = xmlAttrs.getBoolean(R.styleable.PreferenceCategoryEx_dynamic, false);
		state = xmlAttrs.getInt(R.styleable.PreferenceCategoryEx_state, 0);
		xmlAttrs.recycle();
		setLayoutResource(R.layout.preference_category);
	}

	@Override
	public boolean onPrepareAddPreference(Preference preference) {
		preference.onParentChanged(this, shouldDisableDependents());
		return true;
	}

	@Override
	public void onBindViewHolder(PreferenceViewHolder view) {
		super.onBindViewHolder(view);
		TextView title = (TextView) view.findViewById(android.R.id.title);
		title.setText(getTitle() + (unsupported ? " ⨯" : (dynamic ? " ⟲" : "")));
		title.setVisibility((state == 2 || state == 1) ? View.GONE : View.VISIBLE);
		View finalView = view.itemView;
		if (state == 2) {
			finalView.setPadding(childpadding, 0, childpadding, 0);
		}
		else {
			int vertialPadding = getContext().getResources().getDimensionPixelSize(R.dimen.preference_item_padding_top);
			finalView.setPadding(childpadding, vertialPadding, childpadding, vertialPadding);
		}
	}

	public void setUnsupported(boolean value) {
		unsupported = value;
		setEnabled(!value);
	}

	public boolean isDynamic() {
		return dynamic;
	}

	public void hide() {
		state = 2;
		this.notifyChanged();
	}
}

File: app/src/main/java/tv/withaibuild/customiuizer/prefs/PreferenceCategoryEx.kt
package tv.withaibuild.customiuizer.prefs

import android.content.Context
import android.content.res.TypedArray
import android.util.AttributeSet
import android.view.View
import android.widget.TextView
import androidx.preference.Preference
import androidx.preference.PreferenceCategory
import androidx.preference.PreferenceViewHolder
import tv.withaibuild.customiuizer.R

class PreferenceCategoryEx(context: Context, attrs: AttributeSet?) : PreferenceCategory(context, attrs) {

    private val dynamic: Boolean
    private var state = 0 // 0-正常 1-纯区块 2-顶层隐藏
    private var unsupported = false

    init {
        val xmlAttrs: TypedArray = context.obtainStyledAttributes(attrs, R.styleable.PreferenceCategoryEx)
        dynamic = xmlAttrs.getBoolean(R.styleable.PreferenceCategoryEx_dynamic, false)
        state = xmlAttrs.getInt(R.styleable.PreferenceCategoryEx_state, 0)
        xmlAttrs.recycle()
        layoutResource = R.layout.preference_category
    }

    override fun onPrepareAddPreference(preference: Preference): Boolean {
        preference.onParentChanged(this, shouldDisableDependents())
        return true
    }

    override fun onBindViewHolder(holder: PreferenceViewHolder) {
        super.onBindViewHolder(holder)
        val title = holder.findViewById(android.R.id.title) as? TextView
        title?.text = (title?.text?.toString() ?: "") + if (unsupported) " ⨯" else if (dynamic) " ⟲" else ""
        title?.visibility = if (state == 2 || state == 1) View.GONE else View.VISIBLE
        val finalView = holder.itemView
        val childPadding = context.resources.getDimensionPixelSize(R.dimen.preference_item_child_padding)
        if (state == 2) {
            finalView.setPadding(childPadding, 0, childPadding, 0)
        } else {
            val verticalPadding = context.resources.getDimensionPixelSize(R.dimen.preference_item_padding_top)
            finalView.setPadding(childPadding, verticalPadding, childPadding, verticalPadding)
        }
    }

    fun setUnsupported(value: Boolean) {
        unsupported = value
        isEnabled = !value
    }

    fun isDynamic(): Boolean = dynamic

    fun hide() {
        state = 2
        notifyChanged()
    }
}
